test/integration/watch_cache_test.go

```diff
@@ -0,0 +1,113 @@
+package integration
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	"github.com/openshift/origin/pkg/util/restoptions"
+	testutil "github.com/openshift/origin/test/util"
+	testserver "github.com/openshift/origin/test/util/server"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+)
+
+func testWatchCacheWithConfig(t *testing.T, master *configapi.MasterConfig, expectedCacheSize, counterExampleCacheSize int) {
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMasterAPI(master)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	client, err := testutil.GetClusterAdminKubeClient(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// create client with high burst value, otherwise we can only do 5 changes per second
+	config, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	config.Burst = expectedCacheSize
+	patchClient, err := coreclient.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// modify labels of default namespace expectedCacheSize + 1 times
+	defaultNS, err := client.Core().Namespaces().Get("default", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	startVersion, err := strconv.Atoi(defaultNS.ResourceVersion)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	for i := 0; i < expectedCacheSize+10; i++ {
+		defaultNS, err = patchClient.Namespaces().Patch("default", types.StrategicMergePatchType,
+			[]byte(fmt.Sprintf(`{"metadata":{"labels":{"test":"%d"}}}`, i)))
+		if err != nil {
+			t.Fatalf("unexpected patch error: %v", err)
```

Comment:
You're making a high contention field and racing the cache.  I think you'll be ok if you mutate the label too.